pyomo/core/base/set.py

```diff
@@ -1457,17 +1383,145 @@ def clear(self):
+            values.construct()
+        try:
+            val_iter = iter(values)
+        except TypeError:
+            logger.error(
+                "Initializer for Set %s%s returned non-iterable object "
+                "of type %s."
+                % (
+                    self.name,
+                    ("[%s]" % (index,) if self.is_indexed() else ""),
+                    (values if values.__class__ is type else type(values).__name__),
+                )
+            )
+            raise
+
+        if self._dimen is not None:
+            if normalize_index.flatten:
+                val_iter = self._cb_normalized_dimen_verifier(self._dimen, val_iter)
+            else:
+                val_iter = self._cb_raw_dimen_verifier(self._dimen, val_iter)
+        elif normalize_index.flatten:
+            val_iter = map(normalize_index, val_iter)
+        else:
+            val_iter = self._cb_check_set_end(val_iter)
+
+        if self._domain is not Any:
+            val_iter = self._cb_domain_verifier(self._domain, val_iter)
+
+        if self._filter is not None:
+            val_iter = filter(partial(self._filter, self.parent_block()), val_iter)
+
+        if self._validate is not None:
+            val_iter = self._cb_validate(self._validate, self.parent_block(), val_iter)
+
+        # We wrap this check in a try-except because some values
+        #  (like lists) are not hashable and can raise exceptions.
+        try:
+            self._update_impl(val_iter)
+        except Set._SetEndException:
+            pass
 
     def pop(self):
         return self._values.pop()
 
+    def _cb_domain_verifier(self, domain, val_iter):
+        for value in val_iter:
+            if value not in domain:
+                raise ValueError(
+                    "Cannot add value %s to Set %s.\n"
+                    "\tThe value is not in the domain %s"
+                    % (value, self.name, self._domain)
+                )
+            yield value
+
+    def _cb_check_set_end(self, val_iter):
+        for value in val_iter:
+            if value is Set.End:
+                return
+            yield value
+
+    def _cb_validate(self, validate, block, val_iter):
+        for value in val_iter:
+            try:
+                flag = validate(block, value)
+            except:
+                logger.error(
+                    "Exception raised while validating element '%s' "
+                    "for Set %s" % (value, self.name)
+                )
+                raise
+            if not flag:
+                raise ValueError(
+                    "The value=%s violates the validation rule of Set %s"
+                    % (value, self.name)
+                )
+            yield value
+
+    def _cb_normalized_dimen_verifier(self, dimen, val_iter):
+        # It is important that the iterator is an actual iterator
+        val_iter = iter(val_iter)
+        for value in val_iter:
+            if value.__class__ is tuple:
```

Comment:
Mostly because isinstance is slower, and because what we are trying to detect are multi-dimensional indices, which we know (because of normalize_index) is a tuple.